Support `ActiveRecord::Batches::BatchEnumerator`  collections (#409)
Shopify/maintenance_tasks: README.md

My Title,Hello World!
```

### Processing Batch Collections

The Maintenance Tasks gem supports processing Active Records in batches. This
can reduce the number of calls your Task makes to the database. Use
`ActiveRecord::Batches#in_batches` to specify that your Task should process
records in batches. The default batch size is 1000, but a custom size can be
specified.

```ruby
# app/tasks/maintenance/update_posts_in_batches_task.rb
module Maintenance
  class UpdatePostsInBatchesTask < MaintenanceTasks::Task
    def collection
      Post.in_batches
    end

    def count
      Post.count
    end

    def process(batch_of_posts)
      batch_of_posts.update_all(content: "New content added on #{Time.now.utc}")    
    end
  end
end
```

Ensure that you've implemented the following methods:

> * `collection`: return an `ActiveRecord::Batches::BatchEnumerator`.

Reviewer: ```suggestion

* `collection`: return an `ActiveRecord::Batches::BatchEnumerator`.
```

* `process`: do the work of your Task on a batch (`ActiveRecord::Relation`).
* `count`: return the number of rows that will be iterated over (this should

> still be the number of records to be processed, not the number of batches).

Reviewer: Is it? Why not use the number of batches instead? I guess showing the number of records does look nicer.

My Title,Hello World!
```

### Processing Batch Collections

The Maintenance Tasks gem supports processing Active Records in batches. This
can reduce the number of calls your Task makes to the database. Use
`ActiveRecord::Batches#in_batches` to specify that your Task should process
records in batches. The default batch size is 1000, but a custom size can be
specified.

```ruby
# app/tasks/maintenance/update_posts_in_batches_task.rb
module Maintenance
  class UpdatePostsInBatchesTask < MaintenanceTasks::Task
    def collection
      Post.in_batches
    end

    def process(batch_of_posts)

>       batch_of_posts.update_all(content: "New content added on #{Time.now.utc}")    

Reviewer: ```suggestion
      batch_of_posts.update_all(content: "New content added on #{Time.now.utc}")
```

    end
  end
end
```

Ensure that you've implemented the following methods:

* `collection`: return an `ActiveRecord::Batches::BatchEnumerator`.
* `process`: do the work of your Task on a batch (`ActiveRecord::Relation`).

Note that `#count` will be calculated for you based on the number of batches in
your collection, and your Task's progress will be displayed in terms of batches
(not the number of records in the relation).

> **Important!** Batches should only be used if `#process` is performing a batch
> operation such as `#update_all` or `#delete_all`, and should **not** be used
> if you need to iterate over the records in the batch. If you need to iterate
> over individual records, it is more performant to define a collection that
> [directly returns an `ActiveRecord::Relation`](#creating-a-task).

Reviewer: ```suggestion
**Important!** Batches should only be used if `#process` is performing a batch
operation such as `#update_all` or `#delete_all`. If you need to iterate over
individual records, you should define a collection that [directly returns an
`ActiveRecord::Relation`](#creating-a-task). This will also uses batching
internally, but load the records with only one SQL query, whereas Batch
collections will load the primary keys first and then the records, if you call
`each` or any `Enumerable` method.
```

My Title,Hello World!
```

### Processing Batch Collections

The Maintenance Tasks gem supports processing Active Records in batches. This
can reduce the number of calls your Task makes to the database. Use

> `ActiveRecord::Batches#in_batches` to specify that your Task should process
> records in batches. Active Record defaults to 1000 records, but a custom size can be

Reviewer: ```suggestion
`ActiveRecord::Batches#in_batches` on the relation returned by your collection to specify that your Task should process
batches instead of records. Active Record defaults to 1000 records by batch, but a custom size can be
```

specified.

```ruby
# app/tasks/maintenance/update_posts_in_batches_task.rb
module Maintenance
  class UpdatePostsInBatchesTask < MaintenanceTasks::Task
    def collection
      Post.in_batches
    end

    def process(batch_of_posts)
      batch_of_posts.update_all(content: "New content added on #{Time.now.utc}")
    end
  end
end
```

Ensure that you've implemented the following methods:

* `collection`: return an `ActiveRecord::Batches::BatchEnumerator`.
* `process`: do the work of your Task on a batch (`ActiveRecord::Relation`).

Note that `#count` is calculated automatically based on the number of batches in
your collection, and your Task's progress will be displayed in terms of batches
(not the number of records in the relation).

**Important!** Batches should only be used if `#process` is performing a batch
operation such as `#update_all` or `#delete_all`. If you need to iterate over
individual records, you should define a collection that [returns an
`ActiveRecord::Relation`](#creating-a-task). This uses batching
internally, but loads the records with one SQL query. Conversely, batch

> collections load the primary keys first, and then perform a query to load the

Reviewer: ```suggestion
collections load the primary keys of the records of the batch first, and then perform an additional query to load the
```
